Support different landing pages on the same webworker

emord opened the pull request:
A bit of context here: https://docs.google.com/document/d/1YCO2_rLXgfOGABpjmKkaQyje1GX2hLmdAih1tW_CM54/edit#

Basically we're planning to run the REACH pilot on TCL on the same servers as ICDS-CAS, but we can't have ICDS-CAS's custom pages displayed for them. This PR is the first part of that work and will switch the landing page and commcare names based on the url.

There will be some followup commcare-cloud PRs, but I intentionally made this support the current format and new format, so they are not blocking

emord committed:
Add default for CUSTOM_LANDING_TEMPLATE

emord committed:
Add support for login pages per domain

emord committed:
Add support for changing CommCare names based on host

emord committed:
Change URLs in report sidebar to be relative

emord committed:
Add test for new custom landing format

kaapstorm commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -364,8 +365,13 @@ def _login(req, domain_name, template_name):
     custom_landing_page = settings.CUSTOM_LANDING_TEMPLATE
-    if custom_landing_page is not None:
-        template_name = custom_landing_page
+    if custom_landing_page:
```
I can see you're preserving current behaviour by not changing the value of `template_name` that was passed into the function. But the view that allows a non-default value to be set is [`domain_login()`][1] and it is only called in [urls.py here][2], where the template_name is left as default (**unless I'm missing something important**).

If you drop the unused(**‽**) `template_name` param in `login()`, `domain_login()` and`_login()`, and don't bother with setting a default in `CUSTOM_LANDING_TEMPLATE`, then determining the value of `template_name` is simpler:
```python
    custom_landing_page = settings.CUSTOM_LANDING_TEMPLATE
    host = req.get_host()
    if isinstance(custom_landing_page, six.string_types):
        template_name = custom_landing_page
    elif host in custom_landing_page:
        template_name = custom_landing_page[host]
    else:
        template_name = 'login_and_password/login.html'
    if domain_name:
        # ...
```
This also allows a custom landing page AND setting `domain_name` to something other than `None` (for enterprise customers with more than one domain), which the current behaviour does not allow. But maybe checking if we want to allow that is out of scope of this PR.

What do you think?


  [1]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/views.py#L416
  [2]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/urls.py#L65

emord replied:
I removed this unused parameter https://github.com/dimagi/commcare-hq/commit/a53680499faaf80fc56f8c93c07bbb7e92d75af8 but I'm not going to go through with the full change you suggested because that code is written would set the default to our default commcare template which would be undesirable for enterprise envs in a weird case where a proxy change caused the variables set by `get_host()` to be set differently. For ICDS I plan to set it to:

```python
CUSTOM_LANDING_TEMPLATE = {
   "reach.commcarehq.org": "login_and_password/login.html", # likely something more custom later on
   "default": "icds/login.html",
}
```

This prioritizes enterprise server's main sites and ensures that their main sites would never get sent the wrong default landing page

emord replied:
Not to say that this is a great workflow or will last forever, but we've decided that this sort of work is better than maintaining a separate environment on another hosting provider for the next year or longer.

Also I plan on removing the acceptance of string once this work is done and write a changelog for commcare-cloud (echis uses these variables as well, so they will need to be warned about any breaking changes)

kaapstorm replied:
:+1: That makes sense.

kaapstorm reviewed (approved):
Just the one question. I'm happy with this.

emord committed:
Fix hq commcare names for scheduled reports

This test fails without this commit:

corehq.apps.reports.tests.test_scheduled_reports:ScheduledReportSendingTest.test_get_scheduled_report_response

emord committed:
Remove unused default parameter

emord committed:
Fix incorrect call of _login
